### ISSUE TITLE
Support for alternate checksum for a specific source file

### ISSUE THREAD

**Issue by akesandgren** [easybuilders/easybuild-framework#2853]
Since upstream source tar files sometime gets redone and gets a different checksum, it could be useful to have alternative checksums listed for a specific source-tar-file.
This would reduce the problem for those that already have it downloaded when a new EB version comes out and they need to build something again, and would still work for new downloads.

It would also be possible to use in multi_deps exts_lists where the source_tmpl comes from pypi 
with py3 or py2 as part of the name. Something like this could then be used.

```
multi_deps = {'Python': ['3.7.2', '2.7.15']}
exts_list = [
    ('tensorboard', version, {
        'source_tmpl': 'tensorboard-%(version)s-py%(pymajver)s-none-any.whl',
        'source_urls': ['https://pypi.python.org/packages/source/t/tensorboard'],
        'checksums': [('chksum-for-py2-whl', 'chksum-for-py3-whl')],
        'unpack_sources': False,
    }),
]
```

(That last bit would still need a change to how fetch works with multi_deps and exts_lists.)

**Comment by boegel** [comment]
@akesandgren I think there's two different things we can support:

* a checksum mapping, so we can specify checksums for different source files depending on the value of the template `%(pymajver)s`, something like:

```python
'checksums': [{
    'tensorboard-%(version)s-py2-none-any.whl': '...',  # checksum for Python 2 wheel
    'tensorboard-%(version)s-py3-none-any.whl': '...',  # checksum for Python 3 wheel
}],
```

* a tuple of acceptable checksums for a file (similar to what we already support in `sanity_check_paths`), since upstream checksums sometimes change for no good reason (without any actual code changes, see for example https://github.com/easybuilders/easybuild-easyconfigs/pull/8054), where the checksum check passes if *one* of the specified checksums matches

```python
'checksums': [('checksum_one', 'checksum_two')],
```

The concept seems a bit weird, but practice has shown that this sort of thing would be useful in the real world...

**Comment by akesandgren** [comment]
Yeah, looks good to me. (And possibly the combination of the two...)

**Comment by boegel** [comment]
@edmondac has some interest in tackling this, since he needs it to properly handle picking of CUDA sources based on `%(arch)s` (cfr. https://github.com/easybuilders/easybuild-easyconfigs/pull/8136 and https://github.com/easybuilders/easybuild-framework/pull/2841).

The place to implement this would be in the `get_checksum_for` method in `framework/easyblock.py` (see `https://github.com/easybuilders/easybuild-framework/blob/master/easybuild/framework/easyblock.py#L320`).

This already has all the information passed down that you need, it looks like it's just a matter of checking the type of `checksums[index]`, and discriminating between a string value and a `dict` value, where if it's a `dict` value you try to use the `filename` as a key into that dict (and then have proper error handling in case there's another type of value, or if `filename` is not a key in the provided `dict`.

At first sight, `get_checksum_for` is used for sources/patches/extensions, so fixing it there would make it a 'general' fix for all checksums...

The idea of having a list of alternative checksums that are acceptable is also doable, but is a bit of extra work since that involves returning not a single string as return value of `get_checksum_for`, but a tuple of acceptable checksums, which then has to be taken into account everywhere where `get_checksum_for` is used, so that's definitely more painful to implement (it's fine to keep that as a follow-up for later, doesn't have to be tackled in one go).

**Comment by edmondac** [comment]
Dictionary-type checksums are in here: https://github.com/easybuilders/easybuild-framework/pull/2940 , https://github.com/easybuilders/easybuild-easyconfigs/pull/8136 , https://github.com/easybuilders/easybuild-easyblocks/pull/1764

There also seems to be at least some support in the codebase for multiple checksums for a single file already... But not sure how much more would be required (if any) to satisfy this whole ticket.

**Comment by boegel** [comment]
fixed in EasyBuild v3.9.4 thanks to #2946 and #2958, so closing...
